### PR TITLE
Write ME ARCHITECTURE.md section on the migrations table

### PR DESCRIPTION
**Pull request by tomhoule**

### CODE REVIEW

File: migration-engine/ARCHITECTURE.md [hunk 1]
```diff
@@ -26,6 +26,70 @@ Migrate has two main blocks of functionality:
...
+- with `--applied`, mark the existing row as rolled back, and create a new one
+  with `finished_at` == `started_at`. We do this to avoid overwriting the
+  existing record's checksum and timestamps, because that would erase an event
+  that actually happened from the record.
+- with `--rolled-back`, the row's `rolled_back_at` column is populated with the
+  current timestamp.
+
+On `deploy`, Migrate compares the migrations directory on disk with the
+migrations table. If for any row in the migrations table, `finished_at` is null
+and `rolled_|ack_at` is null, it means the migration failed and the failure
```
Comment (Jolg42) [inline migration-engine/ARCHITECTURE.md:79]:
```suggestion
and `rolled_back_at` is null, it means the migration failed and the failure
```

File: migration-engine/ARCHITECTURE.md [hunk 2]
```diff
@@ -26,6 +26,70 @@ Migrate has two main blocks of functionality:
...
+hasn't been resolved yet, so deploy stops there with a detailed error.
+Otherwise, for each migrations in the migrations directory:
+
+- If there is a row in the migrations table,
+    - And `rolled_back_at` is not null: ignore it
+    - And `finished_at` is not null: do not apply the migration
+- If there is no correspondingrow in the migrations table,
```
Comment (Jolg42) [inline migration-engine/ARCHITECTURE.md:86]:
```suggestion
- If there is no corresponding row in the migrations table,
```

**Review by Jolg42 — APPROVED** [review]

File: migration-engine/ARCHITECTURE.md [hunk 3]
```diff
@@ -26,6 +26,70 @@ Migrate has two main blocks of functionality:
...
+## Implementation
+
+### The _prisma_migrations table
+
+We will just call it "the migrations table". It serves the same purpose as the
+migrations tables that nearly all migration tools have. The terminology we use
+is SQL specific, because we only use a migrations table where we have
+migrations, and currently (2022-02), that means SQL connectors.
+
+Prisma's migrations table is a bit more extensive than other tool's tend to be.
```
Comment (tomhoule) [inline migration-engine/ARCHITECTURE.md:38]: tools'

File: migration-engine/ARCHITECTURE.md [hunk 4]
```diff
@@ -26,6 +26,70 @@ Migrate has two main blocks of functionality:
...
+    finished_at             TIMESTAMPTZ,
+    migration_name          VARCHAR(255) NOT NULL,
+    logs                    TEXT,
+    rolled_back_at          TIMESTAMPTZ,
+    started_at              TIMESTAMPTZ NOT NULL DEFAULT now(),
+    applied_steps_count     INTEGER NOT NULL DEFAULT 0
+);
+```
+
+- `id` is a unique random identifier. In practice, a v4 UUID.
```
Comment (tomhoule) [inline migration-engine/ARCHITECTURE.md:55]: random unique

File: migration-engine/ARCHITECTURE.md [hunk 5]
```diff
@@ -26,6 +26,70 @@ Migrate has two main blocks of functionality:
...
+- `checksum` is the sha256 checksum of the migration file. We never override
+  this once it has been written.
+- `finished_at` is the timestamp at which the migration completed. We only
+  write this at the end of a successful migration, so this column being not
+  null means the migration completed without error.
+- `migration_name` is the complete name of the migration directory (without path prefix).
+- `logs` is where we record the error, in case of error.
+- `rolled_back_at` is written by `prisma migrate resolve --rolled-back`, and
```
Comment (tomhoule) [inline migration-engine/ARCHITECTURE.md:63]: delete `--rolled-back`

**Pull request merged**